src/dyncfg/src/lib.rs

```diff
@@ -234,44 +221,118 @@ impl ConfigEntry {
+enum ConfigValAtomic {
     Bool(Arc<AtomicBool>),
-    /// A `u32` shared value.
     U32(Arc<AtomicU32>),
-    /// A `usize` shared value.
     Usize(Arc<AtomicUsize>),
-    /// An `Option<usize>` shared value.
     OptUsize(Arc<RwLock<Option<usize>>>),
-    /// A `String` shared value.
     String(Arc<RwLock<String>>),
-    /// A 'Duration' shared value.
     Duration(Arc<RwLock<Duration>>),
 }
 
+impl From<ConfigVal> for ConfigValAtomic {
+    fn from(val: ConfigVal) -> ConfigValAtomic {
+        match val {
+            ConfigVal::Bool(x) => ConfigValAtomic::Bool(Arc::new(AtomicBool::new(x))),
+            ConfigVal::U32(x) => ConfigValAtomic::U32(Arc::new(AtomicU32::new(x))),
+            ConfigVal::Usize(x) => ConfigValAtomic::Usize(Arc::new(AtomicUsize::new(x))),
+            ConfigVal::OptUsize(x) => ConfigValAtomic::OptUsize(Arc::new(RwLock::new(x))),
+            ConfigVal::String(x) => ConfigValAtomic::String(Arc::new(RwLock::new(x))),
+            ConfigVal::Duration(x) => ConfigValAtomic::Duration(Arc::new(RwLock::new(x))),
+        }
+    }
+}
+
+impl ConfigValAtomic {
+    fn load(&self) -> ConfigVal {
+        match self {
+            ConfigValAtomic::Bool(x) => ConfigVal::Bool(x.load(SeqCst)),
+            ConfigValAtomic::U32(x) => ConfigVal::U32(x.load(SeqCst)),
+            ConfigValAtomic::Usize(x) => ConfigVal::Usize(x.load(SeqCst)),
+            ConfigValAtomic::OptUsize(x) => ConfigVal::OptUsize(*x.read().expect("lock poisoned")),
+            ConfigValAtomic::String(x) => {
+                ConfigVal::String(x.read().expect("lock poisoned").clone())
+            }
+            ConfigValAtomic::Duration(x) => ConfigVal::Duration(*x.read().expect("lock poisoned")),
+        }
+    }
+
+    fn store(&self, val: ConfigVal) {
+        match (self, val) {
+            (ConfigValAtomic::Bool(x), ConfigVal::Bool(val)) => x.store(val, SeqCst),
+            (ConfigValAtomic::U32(x), ConfigVal::U32(val)) => x.store(val, SeqCst),
+            (ConfigValAtomic::Usize(x), ConfigVal::Usize(val)) => x.store(val, SeqCst),
+            (ConfigValAtomic::OptUsize(x), ConfigVal::OptUsize(val)) => {
+                *x.write().expect("lock poisoned") = val
+            }
+            (ConfigValAtomic::String(x), ConfigVal::String(val)) => {
+                *x.write().expect("lock poisoned") = val
+            }
+            (ConfigValAtomic::Duration(x), ConfigVal::Duration(val)) => {
+                *x.write().expect("lock poisoned") = val
+            }
+            (_, val) => panic!("attempted to store {val:?} value in {self:?} parameter"),
+        }
+    }
+}
+
 impl ConfigUpdates {
-    /// Adds the current value of the given config to this set of updates.
+    /// Adds an update for the given config and value.
     ///
     /// If a value of the same config has previously been added to these
     /// updates, replaces it.
-    pub fn add(&mut self, config: &ConfigEntry) {
+    pub fn add<T>(&mut self, config: &Config<T>, val: T)
```

Comment:
nit: when inlining the `where` in the type params doesn't cause it to wrap, I tend to do it (for information density reasons). i.e. `pub fn add<T: ConfigType>(&mut self, config: &Config<T>, val: T) {`

Here and add_dynamic